File: app/src/main/java/com/ga/task/SimpleTaskManager.java
package com.ga.task;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.util.Log;
import android.util.SparseArray;

import junit.framework.Assert;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;

/**
 * Created by alexeyglushkov on 20.09.14.
 */
public class SimpleTaskManager implements TaskManager, TaskPool.TaskPoolListener {

    static final String TAG = "SimpleTaskManager";

    private HandlerThread handlerThread;
    private Handler handler;
    private Handler callbackHandler;
    private TaskExecutor taskExecutor;
    private Object userData;
    private WeakRefList<TaskManagerListener> listeners;

    private TaskPool loadingTasks;
    private TaskProvider waitingTasks;
    private List<TaskProvider> taskProviders; //always sorted by priority

    private SparseArray<Float> limits;
    private SparseArray<Integer> usedSpace; //type -> task count from loadingTasks

    private int maxLoadingTasks;

    public SimpleTaskManager(int maxLoadingTasks) {
        this.taskExecutor = new SimpleTaskExecutor();
        this.maxLoadingTasks = maxLoadingTasks;

        handlerThread = new HandlerThread("SimpleTaskManager Thread");
        handlerThread.start();
        handler = new Handler(handlerThread.getLooper());
        callbackHandler = new Handler(Looper.myLooper());

        loadingTasks = new SimpleTaskPool(handler);
        loadingTasks.addListener(this);

        // TODO: consider putting it to taskProiders list
        waitingTasks = new PriorityTaskProvider(handler, "TaskManagerProviderId");
        waitingTasks.addListener(this);

        listeners = new WeakRefList<TaskManagerListener>();
        taskProviders = new ArrayList<TaskProvider>();

        limits = new SparseArray<Float>();
        usedSpace = new SparseArray<Integer>();
    }

    @Override
    public void setMaxLoadingTasks(int maxLoadingTasks) {
        this.maxLoadingTasks = maxLoadingTasks;
    }

    @Override
    public int getMaxLoadingTasks() {
        return maxLoadingTasks;
    }

    @Override
    public void addTask(final Task task) {
        Assert.assertEquals(task.getTaskStatus(), Task.Status.NotStarted);
        if (!Tasks.isTaskReadyToStart(task)) {
            Log.d(TAG, "Can't put task " + task.getClass().toString() + " because it has been added " + task.getTaskStatus().toString());
            return;
        }

        // TODO: actually I think that problem of adding the same task is not important
        // TaskManager must set Waiting status on the current thread
        task.getPrivate().setTaskStatus(Task.Status.Waiting);

        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                addTaskOnThread(task);
            }
        });
    }

    @Override
    public int getLoadingTaskCount() {
        return loadingTasks.getTaskCount();
    }

    @Override
    public void startImmediately(final Task task) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                if (handleTaskLoadPolicy(task)) {

                    // TODO: write tests for that logic
                    final Task.Callback oldCallBack = task.getTaskCallback();
                    task.setTaskCallback(new Task.Callback() {
                        @Override
                        public void finished(boolean cancelled) {
                            triggerOnTaskRemoved(task, true);

                            if (oldCallBack != null) {
                                oldCallBack.finished(cancelled);
                            }
                        }
                    });

                    triggerOnTaskAdded(task, true);
                    startTaskOnThread(task);
                }
            }
        });
    }

    @Override
    public void cancel(final Task task, final Object info) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                cancelTaskOnThread(task, info);
            }
        });
    }

    @Override
    public void addTaskProvider(final TaskProvider provider) {
        Assert.assertEquals(provider.getHandler(), handler);
        Assert.assertNotNull(provider.getTaskProviderId());

        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                addTaskProviderOnThread(provider);
            }
        });
    }

    public List<TaskProvider> getTaskProviders() {
        checkHandlerThread();

        return taskProviders;
    }

    private void addTaskProviderOnThread(TaskProvider provider) {
        checkHandlerThread();

        TaskProvider oldTaskProvider = getTaskProvider(provider.getTaskProviderId());
        if (oldTaskProvider != null) {
            taskProviders.remove(oldTaskProvider);
        }

        // add in sorted array
        int insertIndex = 0;
        for (int i=0; i<taskProviders.size(); ++i) {
            TaskProvider p = taskProviders.get(i);
            if (provider.getPriority() > p.getPriority()) {
                insertIndex = i;
                break;
            }
        }

        provider.addListener(this); //for snapshots
        taskProviders.add(insertIndex, provider);
    }

    public void setTaskProviderPriority(final TaskProvider provider, final int priority) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                setTaskProviderPriorityOnThread(provider, priority);
            }
        });
    }

    private void setTaskProviderPriorityOnThread(TaskProvider provider, int priority) {
        checkHandlerThread();

        provider.setPriority(priority);
        Collections.sort(taskProviders, new Comparator<TaskProvider>() {
            @Override
            public int compare(TaskProvider lhs, TaskProvider rhs) {
                return Tools.reverseIntCompare(lhs.getPriority(), rhs.getPriority());
            }
        });
    }

    public TaskProvider getTaskProvider(String id) {
        checkHandlerThread();

        for (TaskProvider taskProvider : taskProviders) {
            if (taskProvider.getTaskProviderId().equals(id)) {
                return taskProvider;
            }
        }

        return null;
    }

    @Override
    public void setTaskExecutor(TaskExecutor executor) {
        this.taskExecutor = executor;
    }

    // == TaskPool.TaskPoolListener

    @Override
    public void onTaskAdded(final TaskPool pool, final Task task) {
        checkHandlerThread();

        final boolean isLoadingPool = pool == loadingTasks;
        triggerOnTaskAdded(task, isLoadingPool);

        // run on the next cycle to give a chance to handle added event for other listeners
        // before moving the task to the loading queue
        // otherwise removed event will be sent before added for TaskPool.TaskPoolListener (see functions below)
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (handleTaskLoadPolicy(task)) {
                    checkTasksToRunOnThread();

                } else if (!isLoadingPool) {
                    cancelTaskOnThread(task, null);
                }
            }
        });
    }

    @Override
    public void onTaskRemoved(TaskPool pool, final Task task) {
        checkHandlerThread();

        triggerOnTaskRemoved(task, pool == loadingTasks);
    }

    // == TaskPool interface

    @Override
    public Handler getHandler() {
        return handler;
    }

    @Override
    public void setHandler(Handler handler) {
        /*if (this.handler != null) {
            checkHandlerThread();
        }*/

        this.handler = handler;

        loadingTasks.setHandler(handler);
        waitingTasks.setHandler(handler);
        for (TaskProvider provider : taskProviders) {
            provider.setHandler(handler);
        }
    }

    @Override
    public void removeTask(Task task) {
        cancel(task, null);
    }

    @Override
    public Task getTask(String taskId) {
        checkHandlerThread();

        Task task = loadingTasks.getTask(taskId);

        if (task == null) {
            task = waitingTasks.getTask(taskId);
        }

        if (task == null) {
            for (TaskProvider provider : taskProviders) {
                task = provider.getTask(taskId);
                if (task != null) {
                    break;
                }
            }
        }

        return task;
    }

    @Override
    public int getTaskCount() {
        checkHandlerThread();

        int taskCount = loadingTasks.getTaskCount() + waitingTasks.getTaskCount();
        for (TaskProvider provider : taskProviders) {
            taskCount += provider.getTaskCount();
        }

        return taskCount;
    }

    @Override
    public List<Task> getTasks() {
        List<Task> tasks = new ArrayList<Task>();

        tasks.addAll(loadingTasks.getTasks());
        tasks.addAll(waitingTasks.getTasks());
        for (TaskProvider provider : taskProviders) {
            tasks.addAll(provider.getTasks());
        }

        return tasks;
    }

    @Override
    public void setUserData(Object data) {
        this.userData = data;
    }

    @Override
    public Object getUserData() {
        return userData;
    }

    @Override
    public void removeListener(TaskManagerListener listener) {
        listeners.remove(listener);
    }

    @Override
    public void addListener(TaskManagerListener listener) {
        listeners.add(new WeakReference<TaskManagerListener>(listener));
    }

    @Override
    public void addListener(TaskPoolListener listener) {
        addListener((TaskManagerListener) listener);
    }

    @Override
    public void removeListener(TaskPoolListener listener) {
        removeListener((TaskManagerListener) listener);
    }

    @Override
    public void onTaskStatusChanged(Task task, Task.Status oldStatus, Task.Status newStatus) {
        //unnecessary
    }

    private void triggerOnTaskAdded(Task task, boolean isLoadingQueue) {
        checkHandlerThread();

        for (WeakReference<TaskManagerListener> listener : listeners) {
            listener.get().onTaskAdded(this, task, isLoadingQueue);
        }
    }

    private void triggerOnTaskRemoved(Task task, boolean isLoadingQueue) {
        checkHandlerThread();

        for (WeakReference<TaskManagerListener> listener : listeners) {
            listener.get().onTaskRemoved(this, task, isLoadingQueue);
        }
    }

    // ==

    @Override
    public void removeTaskProvider(final TaskProvider provider) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                removeTaskProviderOnThread(provider);
            }
        });
    }

    private void removeTaskProviderOnThread(TaskProvider provider) {
        checkHandlerThread();

        taskProviders.remove(provider);
    }

    @Override
    public void setLimit(final int taskType, final float availableQueuePart) {
        Tools.runOnHandlerThread(handler, new Runnable() {
            @Override
            public void run() {
                if (availableQueuePart == -1.0f) {
                    limits.remove(taskType);
                } else {
                    limits.put(taskType, availableQueuePart);
                }

                for (WeakReference<TaskManagerListener> listener : listeners) {
                    listener.get().onLimitsChanged(SimpleTaskManager.this, taskType, availableQueuePart);
                }

                /*
                if (snapshot != null) {
                    Tools.runOnHandlerThread(callbackHandler, new Runnable() {
                        @Override
                        public void run() {
                            snapshot.setLoadingLimit(taskType, availableQueuePart);
                        }
                    });
                }*/
            }
        });
    }

    public SparseArray<Float> getLimits() {
        checkHandlerThread();

        return limits.clone();
    }

    @Override
    public SparseArray<Integer> getUsedSpace() {
        return usedSpace.clone();
    }

    public void setWaitingTaskProvider(TaskProvider provider) {
        Assert.assertEquals(provider.getHandler(), handler);

        this.waitingTasks = provider;
    }

    // Private

    private void addTaskOnThread(Task task) {
        checkHandlerThread();

        if (handleTaskLoadPolicy(task)) {
            waitingTasks.addTask(task);
            checkTasksToRunOnThread();
        }
    }

    private boolean handleTaskLoadPolicy(Task task) {
        checkHandlerThread();

        //search for the task with the same id
        if (task.getTaskId() != null) {
            Task addedTask = loadingTasks.getTask(task.getTaskId());
            if (addedTask != null) {
                if (task.getLoadPolicy() == Task.LoadPolicy.CancelAdded) {
                    cancelTaskOnThread(addedTask, null);
                } else {
                    Log.d(TAG, "The task was skipped due to the Load Policy " + task.getLoadPolicy().toString() + task.getClass().toString() + " " + task.getTaskId() + " " + task.getTaskStatus().toString());
                    return false;
                }
            }
        }
        return true;
    }

    private void checkTasksToRunOnThread() {
        checkHandlerThread();

        if (this.loadingTasks.getTaskCount() < this.maxLoadingTasks) {
            final Task task = takeTaskToRunOnThread();

            if (task != null) {
                addLoadingTaskOnThread(task);
                updateUsedSpace(task.getTaskType(), true);
                startTaskOnThread(task);
            }
        }
    }

    private Task takeTaskToRunOnThread() {
        checkHandlerThread();

        List<Integer> taskTypesToFilter = getTaskTypeFilter();
        Task topWaitingTask = this.waitingTasks.getTopTask(taskTypesToFilter);

        int taskProviderIndex = -1;
        Task topPriorityTask = null;
        int topPriority = -1;

        if (topWaitingTask != null) {
            topPriorityTask = topWaitingTask;
            topPriority = topWaitingTask.getTaskPriority();
        }

        int i = 0;
        for (TaskProvider provider : taskProviders) {
            if (provider != null) {
                Task t = provider.getTopTask(taskTypesToFilter);
                if ( t != null && t.getTaskPriority() > topPriority) {
                    topPriorityTask = t;
                    taskProviderIndex = i;
                }
            }

            ++i;
        }

        if (topPriorityTask != null && !reachedLimit(topPriorityTask.getTaskType())) {
            if (taskProviderIndex == -1 && topPriorityTask != null) {
                topPriorityTask = waitingTasks.takeTopTask(taskTypesToFilter);

            } else if (taskProviderIndex != -1) {
                topPriorityTask = taskProviders.get(taskProviderIndex).takeTopTask(taskTypesToFilter);
            }
        } else {
            topPriorityTask = null;
        }

        return topPriorityTask;
    }

    private List<Integer> getTaskTypeFilter() {
        List<Integer> taskTypesToFilter = new ArrayList<Integer>();
        for (int i = 0; i < limits.size(); i++) {
            if (reachedLimit(limits.keyAt(i))) {
                taskTypesToFilter.add(limits.keyAt(i));
            }
        }
        return taskTypesToFilter;
    }

    private void startTaskOnThread(final Task task) {
        checkHandlerThread();
        Assert.assertTrue(Tasks.isTaskReadyToStart(task));

        logTask(task, "Task started");
        task.getPrivate().setTaskStatus(Task.Status.Started);
        task.getPrivate().setTaskStartDate(new Date());

        final Task.Callback originalCallback = task.getTaskCallback();
        Task.Callback callback = new Task.Callback() {
            @Override
            public void finished(final boolean cancelled) {
                Tools.runOnHandlerThread(handler, new Runnable() {
                    @Override
                    public void run() {
                        logTask(task, "Task finished");
                        updateUsedSpace(task.getTaskType(), false);

                        Task.Status newStatus = cancelled ? Task.Status.Cancelled : Task.Status.Finished;
                        handleTaskCompletionOnThread(task, originalCallback, newStatus);
                    }
                });
            }
        };
        task.setTaskCallback(callback);
        taskExecutor.executeTask(task);
    }

    private void logTask(Task task, String prefix) {
        Log.d(TAG, prefix + " " + task.getClass().toString() + "(" + task.getTaskStatus() + ")" + " id= " + task.getTaskId() + " priority= " + task.getTaskPriority() + " time " + task.getTaskDuration());
    }

    private void checkHandlerThread() {
        Assert.assertEquals(Looper.myLooper(), handler.getLooper());
    }

    public void handleTaskCompletionOnThread(final Task task, final Task.Callback callback, final Task.Status status) {
        checkHandlerThread();
        task.getPrivate().setTaskStatus(status);
        task.getPrivate().clearAllListeners();

        Tools.runOnHandlerThread(callbackHandler, new Runnable() {
            @Override
            public void run() {
                if (callback != null) {
                    callback.finished(status == Task.Status.Cancelled);
                }
            }
        });

        //TODO: for the another status like cancelled new task won't be started
        //but we cant't just call checkTasksToRunOnThread due to Task.LoadPolicy.CancelAdded
        //because we want to have new task already in waiting queue but now it isn't
        if (status == Task.Status.Finished) {
            checkTasksToRunOnThread();
        }
    }

    void cancelTaskOnThread(Task task, final Object info) {
        checkHandlerThread();

        if (!task.getPrivate().getNeedCancelTask()) {
            Task.Status st = task.getTaskStatus();
            task.getPrivate().cancelTask(info);

            if (st == Task.Status.Waiting) {
                // the task will be removed from providers automatically
                handleTaskCompletionOnThread(task, task.getTaskCallback(), Task.Status.Cancelled);

                logTask(task, "Cancelled");
            }
        }
    }

    private boolean reachedLimit(int taskType) {
        if (limits.get(taskType,-1.0f) == -1.0f) {
            return false;
        }

        return (float)usedSpace.get(taskType,0) / (float)maxLoadingTasks >= limits.get(taskType, 0.0f);
    }

    private void updateUsedSpace(int taskType, boolean add) {
        Integer count = usedSpace.get(taskType, 0);
        if (add) {
            ++count;
        } else {
            Assert.assertTrue(count > 0);
            --count;
        }

        usedSpace.put(taskType, count);
    }

    // helpers

    void addLoadingTaskOnThread(Task task) {
        checkHandlerThread();

        this.loadingTasks.addTask(task);
        Log.d(TAG, "loading task count " + loadingTasks.getTaskCount());
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/SimpleTaskManager.kt
package com.example.alexeyglushkov.taskmanager.task

import android.os.Handler
import android.os.HandlerThread
import android.os.Looper

import android.util.Log
import androidx.annotation.WorkerThread
import com.example.alexeyglushkov.streamlib.progress.ProgressListener

import com.example.alexeyglushkov.tools.CancelError
import com.example.alexeyglushkov.tools.HandlerTools

import org.junit.Assert

import java.lang.ref.WeakReference
import java.util.ArrayList
import java.util.Comparator
import java.util.Date

import kotlinx.coroutines.*
import kotlinx.coroutines.android.asCoroutineDispatcher

/**
 * Created by alexeyglushkov on 20.09.14.
 */
open class SimpleTaskManager : TaskManager, TaskPool.Listener {
    companion object {
        internal val TAG = "SimpleTaskManager"
        const val WaitingTaskProviderId = "WaitingTaskProviderId"
    }

    var _threadRunner: ThreadRunner = InstantThreadRunner()
    // TODO: try to use channel instead of a single thread
    override var threadRunner: ThreadRunner
        get() = _threadRunner
        set(value) {
            _threadRunner = value

            loadingTasks.threadRunner = value
            for (provider in _taskProviders) {
                provider.threadRunner = value
            }
        }

    private lateinit var callbackHandler: Handler
    override var userData: Any? = null
    private var listeners = WeakRefList<TaskManager.Listener>()

    private var _taskScope: CoroutineScope? = null
    private var taskScope: CoroutineScope
        get() = _taskScope!!
        set(value) {
            _taskScope = value
        }

    private lateinit var loadingTasks: TaskPool
    private lateinit var waitingTasks: TaskProvider

    private var taskToJobMap = HashMap<Task, Job>()

    // TODO: check out how we can avoid using SafeList or another solution
    // TODO: think about weakref
    private lateinit var _taskProviders: SafeList<TaskProvider>
    override val taskProviders: List<TaskProvider>
         get() {
            return if (threadRunner.isOnThread())
                _taskProviders.originalList
            else
                _taskProviders.safeList
        }

    private lateinit var _coordinator: TaskManagerCoordinator
    override var taskManagerCoordinator: TaskManagerCoordinator
        set(value) {
            // it's experimental
            // TODO: need to handle it properly
            setupCoordinator(value)
        }
        get() = _coordinator

    override fun getLoadingTaskCount(): Int {
        return loadingTasks.getTaskCount()
    }

    @WorkerThread
    override fun getTaskCount(): Int {
        return threadRunner.run {
            var taskCount = loadingTasks.getTaskCount()
            for (provider in _taskProviders) {
                taskCount += provider.getTaskCount()
            }

            return@run taskCount
        }
    }

    @WorkerThread
    override fun getTasks(): List<Task> {
        val tasks = ArrayList<Task>()

        tasks.addAll(loadingTasks.getTasks())
        for (provider in _taskProviders) {
            tasks.addAll(provider.getTasks())
        }

        return tasks
    }

    constructor(inCoordinator: TaskManagerCoordinator) {
        init(inCoordinator, null, null)
    }

    constructor(inCoordinator: TaskManagerCoordinator, scope: CoroutineScope, taskScope: CoroutineScope) {
        init(inCoordinator, scope, taskScope)
    }

    private fun init(inCoordinator: TaskManagerCoordinator, inScope: CoroutineScope?, inTaskScope: CoroutineScope?) {
        initScope(inScope)
        initTaskSope(inTaskScope)
        setupCoordinator(inCoordinator)

        callbackHandler = Handler(Looper.myLooper())

        loadingTasks = ListTaskPool(threadRunner)
        loadingTasks.addListener(this)

        _taskProviders = createTaskProviders()
        createWaitingTaskProvider()
    }

    private fun initScope(inScope: CoroutineScope?) {
        val scope = if (inScope != null) {
            inScope
        } else {
            val localHandlerThread = HandlerThread("SimpleTaskManager Thread")
            localHandlerThread.start()
            val handler = Handler(localHandlerThread.looper)
            val dispatcher = handler.asCoroutineDispatcher("SimpleTaskManager handler dispatcher")
            CoroutineScope(dispatcher + SupervisorJob())
        }

        _threadRunner = ScopeThreadRunner(scope, "SimpleTaskManagerScopeTheadId")
        scope.launch {
            threadRunner.setup()
        }
    }

    private fun setupCoordinator(aCooridinator: TaskManagerCoordinator) {
        _coordinator = aCooridinator
        _coordinator.threadRunner = threadRunner
    }

    private fun initTaskSope(inScope: CoroutineScope?) {
        taskScope = inScope ?: CoroutineScope(Dispatchers.IO + SupervisorJob())
    }

    private fun createTaskProviders(): SafeList<TaskProvider> {
        val sortedList = SortedList(Comparator<TaskProvider> { lhs, rhs ->
            if (lhs.priority == rhs.priority) {
                return@Comparator 0
            }

            if (lhs.priority > rhs.priority) {
                -1
            } else 1
        })

        return SafeList(sortedList, callbackHandler)
    }

    private fun createWaitingTaskProvider() {
        waitingTasks = PriorityTaskProvider(threadRunner, WaitingTaskProviderId)
        waitingTasks.addListener(this)

        _taskProviders.add(waitingTasks)
    }

    override fun addTask(task: Task) {
        if (task !is TaskBase) { assert(false); return }

        // TODO: think how to handle adding two same task
        // Assert.assertEquals(task.getTaskStatus(), Task.Status.NotStarted);
        if (!Tasks.isTaskReadyToStart(task)) {
            Log.d(TAG, "addTask: Can't add task " + task.javaClass.toString() + " because it has been added " + task.taskStatus.toString())
            return
        }

        // TODO: actually I think that problem of adding the same task is not important
        // TaskManager must set Waiting status on the current thread
        setTaskStatus(task, Task.Status.Waiting)

        threadRunner.launch {
            // task will be launched in onTaskAdded method
            waitingTasks.addTask(task)
        }
    }

    override fun startImmediately(task: Task) {
        threadRunner.launchSuspend {
            if (handleTaskLoadPolicy(task, null)) {
                startTaskOnThread(task)
            }
        }
    }

    override fun cancel(task: Task, info: Any?) {
        threadRunner.launch {
            cancelTaskOnThread(task, info)
        }
    }

    override fun addTaskProvider(provider: TaskProvider) {
        Assert.assertEquals(provider.threadRunner, threadRunner)
        Assert.assertNotNull(provider.taskProviderId)

        threadRunner.launch {
            addTaskProviderOnThread(provider)
        }
    }

    @WorkerThread
    private fun addTaskProviderOnThread(provider: TaskProvider) {
        threadRunner.checkThread()

        val oldTaskProvider = getTaskProvider(provider.taskProviderId)
        if (oldTaskProvider != null) {
            removeTaskProviderOnThread(oldTaskProvider)
        }

        provider.addListener(this)
        _taskProviders.add(provider)
        provider.taskFilter = taskManagerCoordinator.taskFilter
    }

    override fun setTaskProviderPriority(provider: TaskProvider, priority: Int) {
        threadRunner.launch {
            setTaskProviderPriorityOnThread(provider, priority)
        }
    }

    @WorkerThread
    private fun setTaskProviderPriorityOnThread(provider: TaskProvider, priority: Int) {
        threadRunner.checkThread()

        provider.priority = priority
        (_taskProviders.originalList as SortedList<*>).updateSortedOrder()
    }

    override fun getTaskProvider(id: String): TaskProvider? {
        return findProvider(taskProviders, id)
    }

    private fun findProvider(providers: List<TaskProvider>, id: String): TaskProvider? {
        for (taskProvider in providers) {
            if (taskProvider.taskProviderId == id) {
                return taskProvider
            }
        }
        return null
    }

    // == TaskPool.TaskPoolListener

    @WorkerThread
    override fun onTaskConflict(pool: TaskPool, newTask: Task, oldTask: Task): Task {
        threadRunner.checkThread()
        Assert.assertTrue(pool != loadingTasks)

        return threadRunner.run {
            resolveTaskConflict(newTask, oldTask)
        }
    }

    @WorkerThread
    override fun onTaskAdded(pool: TaskPool, task: Task) {
        threadRunner.launch {
            val isLoadingPool = pool === loadingTasks
            if (isLoadingPool) {
                taskManagerCoordinator.onTaskStartedLoading(pool, task)
            }

            logTask(task, "onTaskAdded $isLoadingPool")
            triggerOnTaskAddedOnThread(task, isLoadingPool)

            if (!isLoadingPool) {
                if (handleTaskLoadPolicy(task, pool)) {
                    checkTasksToRunOnThread()
                }
            }
        }
    }

    @WorkerThread
    override fun onTaskRemoved(pool: TaskPool, task: Task) {
        threadRunner.launch {
            val isLoadingPool = pool === loadingTasks
            if (isLoadingPool) {
                taskManagerCoordinator.onTaskFinishedLoading(pool, task)
            }

            logTask(task, "onTaskRemoved $isLoadingPool")
            triggerOnTaskRemovedOnThread(task, isLoadingPool)
        }
    }

    @WorkerThread
    override fun onTaskCancelled(pool: TaskPool, task: Task, info: Any?) {
        threadRunner.launch {
            cancelTaskOnThread(task, info)
        }
    }

    // == TaskPool interface

    override fun cancelTask(task: Task, info: Any?) {
        cancel(task, info)
    }

    override fun removeTask(task: Task) {
        cancel(task, null)
    }

    @WorkerThread
    override fun getTask(taskId: String): Task? {
        threadRunner.checkThread()
        return threadRunner.run {
            var task: Task? = loadingTasks.getTask(taskId)
            if (task == null) {
                for (provider in _taskProviders) {
                    task = provider.getTask(taskId)
                    if (task != null) {
                        break
                    }
                }
            }

            return@run task
        }
    }

    override fun removeListener(listener: TaskManager.Listener) {
        listeners.removeValue(listener)
    }

    override fun addListener(listener: TaskManager.Listener) {
        listeners.add(WeakReference(listener))
    }

    override fun addListener(listener: TaskPool.Listener) {
        //unnecessary
    }

    override fun removeListener(listener: TaskPool.Listener) {
        //unnecessary
    }

    override fun onTaskStatusChanged(task: Task, oldStatus: Task.Status, newStatus: Task.Status) {
        //unnecessary
    }

    @WorkerThread
    private fun triggerOnTaskAddedOnThread(task: Task, isLoadingQueue: Boolean) {
        threadRunner.checkThread()

        for (listener in listeners) {
            listener.get()?.onTaskAdded(this, task, isLoadingQueue)
        }
    }

    @WorkerThread
    private fun triggerOnTaskRemovedOnThread(task: Task, isLoadingQueue: Boolean) {
        threadRunner.checkThread()

        for (listener in listeners) {
            listener.get()?.onTaskRemoved(this, task, isLoadingQueue)
        }
    }

    @WorkerThread
    private fun triggerOnTaskStatusChangedOnThread(task: Task, oldStatus: Task.Status) {
        threadRunner.checkThread()

        for (listener in listeners) {
            listener.get()?.onTaskStatusChanged(task, oldStatus)
        }
    }

    // ==

    override fun removeTaskProvider(provider: TaskProvider) {
        threadRunner.launch {
            removeTaskProviderOnThread(provider)
        }
    }

    @WorkerThread
    private fun removeTaskProviderOnThread(provider: TaskProvider) {
        threadRunner.checkThread()

        // cancel all tasks
        for (task in ArrayList(provider.getTasks())) {
            cancelTaskOnThread(task, null)
        }

        _taskProviders.remove(provider)
        provider.taskFilter = null
    }

    fun setWaitingTaskProvider(provider: TaskProvider) {
        Assert.assertEquals(provider.threadRunner, threadRunner)

        provider.taskProviderId = WaitingTaskProviderId
        addTaskProvider(provider)
    }

    // Private

    // returns true when load policy checks pass
    @WorkerThread
    private fun handleTaskLoadPolicy(newTask: Task, taskPool: TaskPool?): Boolean {
        threadRunner.checkThread()
        if (Tasks.isTaskBlocked(newTask)) {
            return false
        }

        val taskId = newTask.taskId
        if (taskId != null) {
            val pools = ArrayList<TaskPool>(taskProviders) + loadingTasks
            for (pool in pools) {
                if (pool != taskPool) {
                    val oldTask = pool.getTask(taskId)
                    if (oldTask != null) {
                        assert(oldTask != newTask)
                        val resultTask = resolveTaskConflict(newTask, oldTask)
                        if (resultTask == oldTask) {
                            return false // quit when task fails
                        }
                    }
                }

                if (Tasks.isTaskBlocked(newTask)) {
                    return false
                }
            }
        }

        return true
    }

    // returns newTask if we allow adding it in a provider
    @WorkerThread
    private fun resolveTaskConflict(newTask: Task, currentTask: Task): Task {
        threadRunner.checkThread()

        return when (newTask.loadPolicy) {
            Task.LoadPolicy.AddDependencyIfAlreadyAdded -> {
                logTask(newTask, "Conflict: wait until the current task finishes")
                logTask(currentTask, "Conflict: the current task")
                // TODO: we need to call addTaskDependency for all other tasks with this id or probably create a list to store all such dependant tasks
                newTask.addTaskDependency(currentTask)
                newTask
            }
            Task.LoadPolicy.CompleteWhenAlreadyAddedCompletes -> {
                logTask(newTask, "Conflict: this task will complete with the current task")
                logTask(currentTask, "Conflict: the current task")
                connectTaskCompletions(newTask, currentTask)
                newTask
            }
            Task.LoadPolicy.CancelPreviouslyAdded -> {
                logTask(currentTask, "Conflict: this current task is cancelled")
                logTask(newTask, "Conflict: the new task")
                cancelTaskOnThread(currentTask, null)
                newTask
            }
            Task.LoadPolicy.SkipIfAlreadyAdded -> {
                logTask(newTask, "Conflict: this task was skipped as there is another task already")
                logTask(currentTask, "Conflict: another task")
                cancelTaskOnThread(newTask, null)
                currentTask
            }
        }
    }

    // completes aTask when toTask completes with setting taskResult
    // syncs progress state too
    @WorkerThread
    private fun connectTaskCompletions(newTask: Task, currentTask: Task) {
        addTaskBlockedDependency(newTask, currentTask)

        currentTask.addTaskProgressListener(ProgressListener { _, progressInfo ->
            progressInfo?.let {
                newTask as TaskBase
                newTask.private.triggerProgressListeners(progressInfo)
            }
        })

        currentTask.addTaskStatusListener(object : Task.StatusListener {
            override fun onTaskStatusChanged(task: Task, oldStatus: Task.Status, newStatus: Task.Status) {
                if (Tasks.isTaskFinished(task)) {
                    currentTask.removeTaskStatusListener(this)

                    if (task.taskStatus != Task.Status.Cancelled) {
                        newTask as TaskBase
                        newTask.private.taskResult = currentTask.taskResult
                        handleTaskFinishOnThread(newTask, false)
                    } else {
                        setTaskStatus(newTask, Task.Status.Waiting)
                    }
                }
            }
        })
    }

    @WorkerThread
    private fun addTaskBlockedDependency(newTask: Task, currentTask: Task) {
        setTaskStatus(newTask, Task.Status.Blocked)
        newTask.addTaskDependency(currentTask)
    }

    private fun setTaskStatus(task: Task, status: Task.Status) {
        val oldStatus = task.taskStatus
        task as TaskBase
        task.private.taskStatus = status

        threadRunner.launch {
            triggerOnTaskStatusChangedOnThread(task, oldStatus)
        }
    }

    @WorkerThread
    private fun checkTasksToRunOnThread() {
        threadRunner.checkThread()
        threadRunner.launchSuspend {
            if (taskManagerCoordinator.canAddMoreTasks()) {
                val task = takeTaskToRunOnThread()

                if (task != null) {
                    logTask(task, "Have taken task")
                    startTaskOnThread(task)
                }
            }
        }
    }

    @WorkerThread
    private fun takeTaskToRunOnThread(): Task? {
        threadRunner.checkThread()

        var topTaskProvider: TaskProvider? = null
        var topPriorityTask: Task? = null
        var topPriority = -1

        for (provider in _taskProviders) {
            val t = provider.getTopTask()
            if (t != null && t.taskPriority > topPriority) {
                topPriorityTask = t
                topPriority = t.taskPriority
                topTaskProvider = provider
            }
        }

        return if (topTaskProvider != null && topPriorityTask != null) {
            topTaskProvider.takeTopTask()
        } else {
            null
        }
    }

    @WorkerThread
    private suspend fun startTaskOnThread(task: Task) {
        if (task !is TaskBase) { assert(false); return }
        threadRunner.checkThread()
        Assert.assertTrue(Tasks.isTaskReadyToStart(task))

        addLoadingTaskOnThread(task)

        logTask(task, "Task started")
        setTaskStatus(task, Task.Status.Started)
        task.private.setTaskStartDate(Date())

        val job = SupervisorJob()
        taskToJobMap[task] = job

        var isCancelled = false
        withContext(taskScope.coroutineContext + job) {
            try {
                task.startTask()
            } catch (e: Throwable) {
                isCancelled = job.isCancelled
            }
        }

        logTask(task, "Task onCompleted" + if (isCancelled) " (Cancelled)" else "")
        taskToJobMap.remove(task)
        if (!isCancelled) {
            handleTaskFinishOnThread(task, false)
        }
    }

    @WorkerThread
    private fun addLoadingTaskOnThread(task: Task) {
        threadRunner.checkThread()
        loadingTasks.addTask(task)
        Log.d(TAG, "loading task count " + loadingTasks.getTaskCount())
    }

    @WorkerThread
    private fun handleTaskFinishOnThread(task: Task, isCancelled: Boolean) {
        if (task !is TaskBase) { assert(false); return }
        threadRunner.checkThread()

        val status = if (isCancelled) Task.Status.Cancelled else Task.Status.Completed
        if (isCancelled) {
            task.private.taskError = CancelError()
        }

        // the task will be removed from the provider automatically
        logTask(task, "finished")
        setTaskStatus(task, status)
        task.private.clearAllListeners()

        // TODO: use callback scope
        task.finishCallback?.let { finishCallback ->
            HandlerTools.runOnHandlerThread(callbackHandler) {
                finishCallback.onCompleted(isCancelled)
            }
        }

        //TODO: for another status like cancelled new task won't be started
        //but we cant't just call checkTasksToRunOnThread because of Task.LoadPolicy.CancelAdded
        //because we want to have new task already in waiting queue but now it isn't
        if (task.taskStatus == Task.Status.Completed) {
            checkTasksToRunOnThread()
        }
    }

    @WorkerThread
    private fun cancelTaskOnThread(task: Task, info: Any?) {
        if (task !is TaskBase) { assert(false); return }
        threadRunner.checkThread()

        if (!task.private.needCancelTask && !Tasks.isTaskFinished(task)) {
            val st = task.taskStatus
            task.private.cancelTask(info)

            val canBeCancelledImmediately = task.private.canBeCancelledImmediately()
            if (Tasks.isTaskReadyToStart(task)) {
                handleTaskFinishOnThread(task, true)
                logTask(task, "Cancelled")

            } else if (canBeCancelledImmediately) {
                if (st == Task.Status.Started) {
                    if (canBeCancelledImmediately) {
                        handleTaskFinishOnThread(task, true)
                        logTask(task, "Immediately Cancelled")

                        taskToJobMap[task]?.cancel()
                    } // else wait until the task handles needCancelTask
                } // else ignore, the callback is already called
            }
        }
    }

    /// Helpers

    private fun logTask(task: Task, prefix: String) {
        Tasks.logTask(TAG, task, prefix)
    }
}
